tests/virt_control_plane_test.go

```diff
@@ -0,0 +1,111 @@
+/*
+ * This file is part of the KubeVirt project
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * Copyright 2019 Red Hat, Inc.
+ *
+ */
+
+package tests_test
+
+import (
+	"fmt"
+
+	. "github.com/onsi/ginkgo"
+	. "github.com/onsi/gomega"
+
+	"kubevirt.io/kubevirt/tests"
+)
+
+const (
+	DefaultTimeout = "--timeout=60s"
+)
+
+var _ = Describe("KubeVirt control plane resilience", func() {
+
+	var err error
+
+	tests.FlagParse()
+
+	BeforeEach(func() {
+		tests.SkipIfNoCmd("kubectl")
+		tests.BeforeTestCleanup()
+	})
+
+	runCommandOnNode := func(command string, node string, args ...string) (err error) {
+		cmdName := tests.GetK8sCmdClient()
+		newArgs := make([]string, 0)
+		if tests.IsOpenShift() {
+			// if the cluster is openshift we need to append `adm` for the commands `drain` and `uncordon`
+			// as the oc binary is used
+			newArgs = append(newArgs, "adm")
+		}
+		newArgs = append(newArgs, command)
+		newArgs = append(newArgs, node)
+		newArgs = append(newArgs, args...)
+		_, _, err = tests.RunCommandWithNS("", cmdName, newArgs...)
+		return
+	}
+
+	uncordonNode := func(node string) (err error) {
+		err = runCommandOnNode("uncordon", node)
+		return
+	}
+
+	AfterEach(func() {
+		err = uncordonNode("node01")
```

Comment:
You are not supposed to make assumptions about the nodes. The tests will then only work on kubevirtci and not in general.

A few suggestions:

 1) If a test requires more than one node, you have to skip the test if only one node is there. We do that normally like this:

```
		nodes := tests.GetAllSchedulableNodes(virtClient)
		Expect(nodes.Items).ToNot(BeEmpty(), "There should be some compute node")

		if len(nodes.Items) < 2 {
			Skip("Migration tests require at least 2 nodes")
		}
```

Hape if you make something like this out of this:

```
SkipIfSchedulableNodesBelow(2)
```

2) in order to not depend on the count of available nodes, you will have to be a little creative:
 * Get a schedulable node
 * Add a nodeSelector for that node to our deployment
 * Wait until the pod is running on that node
 * drain it

Reply:
@rmohr thanks for your input, great points, will integrate a more flexible setup concerning the number of nodes. IIUC I should restructure the test as to expect a drain on the _last node_ to fail (because of the PDB) _iif_ all other nodes are cordoned. I saw something concerning iterating on present nodes in `tests.BeforeTestCleanup()`, will adapt on this.

Regarding the number of nodes: the PDB is about safeguarding a pod for voluntary evictions, so a test with one node wouldn't make sense, right? I mean in that case there would be manually intervention required by the admin anyway.

Reply:
> IIUC I should restructure the test as to expect a drain on the _last node_ to fail (because of the PDB) _iif_ all other nodes are cordoned.

You don't have to cordon anything. Set a nodeSelector on the Deployment which selects the node where the pods are currently running on. Then the pods can't bescheduled to any other node. Therefore any `evict` should fail.

Reply:
If you would cordon all nodes that would be fun if you have more nodes :D

Reply:
@rmohr if you have time it'd be great if you could have another look